torch/csrc/jit/mobile/type_parser.cpp

```diff
@@ -164,23 +167,190 @@ class TypeParser {
+        super(Bar, self).__init__()
+        self.foo = Foo(torch.tensor(1))
+
+    def forward(self, a: torch.Tensor):
+        self.foo = Foo(a)
+        return self.foo
+
+The corresponding bytecode will be:
+(6,
+ ('__torch__.Bar.forward',
+  (('instructions',
+    (('STOREN', 1, 2),
+     ('LOAD', 1, 0),
+     ('MOVE', 2, 0),
+     ('NAMED_TUPLE_CONSTRUCT', 0, 1),
+     ('SET_ATTR', 2, 0),
+     ('MOVE', 1, 0),
+     ('GET_ATTR', 2, 0),
+     ('RET', 0, 0))),
+   ('operators', ()),
+   ('constants', ()),
+   ('types', (('__torch__.Foo', ('NamedTuple', (('id', 'Tensor'),))),)),
+   ('register_size', 2)),
+  (('arguments',
+    ((('name', 'self'), ('type', '__torch__.Bar'), ('default_value', None)),
+     (('name', 'a'), ('type', 'Tensor'), ('default_value', None)))),
+   ('returns',
+    ((('name', ''), ('type', 'Tuple[Tensor]'), ('default_value', None)),)))))
+
+*/
+class CustomTypeParser final {
+ public:
+  CustomTypeParser(CustomTypeParser const&) = delete;
+  CustomTypeParser& operator=(CustomTypeParser const&) = delete;
+  CustomTypeParser() {
+    registerParserFunction(kTypingNamedTuple, parseCustomTypeNamedTuple);
+  }
+
+  TypePtr parse(IValue custom_type) {
+    // Parse custom type, currently only support NamedTuple
+    if (custom_type.isTuple()) {
+      auto tuple_difinition = custom_type.toTuple()->elements();
+      const std::string type_name = getType(custom_type);
+      if (hasParserFunction(type_name)) {
+        return parserFunctions()[type_name](custom_type);
+      } else {
+        TORCH_CHECK(
+            false,
+            "The custom type: ",
+            type_name,
+            " is currently not supported.")
+      }
+    }
+    TORCH_CHECK(
+        false,
+        "For custom type, the definition needs to tuple, however receives : ",
+        custom_type.tagKind())
+    return nullptr;
+  }
+
+  [[nodiscard]] std::unordered_set<std::string> getSupportedTypes() {
+    std::unordered_set<std::string> supported_type_list;
+    for (const auto& type_name : parserFunctions()) {
+      supported_type_list.insert(type_name.first);
+    }
+    return supported_type_list;
+  }
+
+ private:
+  [[nodiscard]] bool hasParserFunction(const std::string& custom_type) const {
+    return parserFunctions().count(custom_type);
+  }
+
+  std::string getType(IValue custom_type) {
+    auto difinition = custom_type.toTuple()->elements()[1];
+    return difinition.toTuple()->elements()[0].toString()->string();
+  }
+
+  [[nodiscard]] std::
+      unordered_map<std::string, std::function<TypePtr(IValue&)>>&
+      parserFunctions() const {
+    static std::unordered_map<std::string, std::function<TypePtr(IValue&)>>
+        custom_type_parser_functions;
+    return custom_type_parser_functions;
+  }
+
+  void registerParserFunction(
+      const std::string& custom_type,
+      const std::function<TypePtr(IValue&)>& custom_type_parser_function) {
+    parserFunctions()[custom_type] = custom_type_parser_function;
```

Comment:
If it's registered, it would make more sense to put it into a separate cpp file, so that it can be optionally built into the binary.

Reply:
Not applicable to the new implementation.